core/admin/mailu/utils.py

```diff
@@ -48,3 +63,79 @@ def init_app(self, app):
 
 # Data migrate
 migrate = flask_migrate.Migrate()
+
+
+# session store
+class NullSigner(object):
+    """NullSigner does not sign nor unsign"""
+    def __init__(self, *args, **kwargs):
+        pass
+    def sign(self, value):
+        """Signs the given string."""
+        return want_bytes(value)
+    def unsign(self, signed_value):
+        """Unsigns the given string."""
+        return want_bytes(signed_value)
+
+class KVSessionIntf(flask_kvsession.KVSessionInterface):
+    """ KVSession interface allowing to run int function on first access """
+    def __init__(self, app, init_fn=None):
+        if init_fn:
+            app.kvsession_init = init_fn
+        else:
+            self._first_run(None)
+    def _first_run(self, app):
+        if app:
+            app.kvsession_init()
+        self.open_session = super().open_session
+        self.save_session = super().save_session
+    def open_session(self, app, request):
+        self._first_run(app)
+        return super().open_session(app, request)
+    def save_session(self, app, session, response):
+        self._first_run(app)
+        return super().save_session(app, session, response)
+
+class KVSessionExt(flask_kvsession.KVSessionExtension):
+    """ Activates Flask-KVSession for an application. """
+    def init_kvstore(self, config):
+        """ Initialize kvstore - fallback to DictStore without REDIS_ADDRESS """
+        if addr := config.get('REDIS_ADDRESS'):
+            self.default_kvstore = RedisStore(redis.StrictRedis().from_url(f'redis://{addr}/3'))
+        else:
+            self.default_kvstore = DictStore()
+
+    def cleanup_sessions(self, app=None, dkey=None, dvalue=None):
+        """ Remove sessions from the store. """
+        if not app:
+            app = flask.current_app
+        if dkey is None and dvalue is None:
+            now = datetime.utcnow()
+            for key in app.kvsession_store.keys():
+                try:
+                    sid = flask_kvsession.SessionID.unserialize(key)
+                except ValueError:
+                    pass
+                else:
+                    if sid.has_expired(
+                        app.config['PERMANENT_SESSION_LIFETIME'],
+                        now
+                    ):
+                        app.kvsession_store.delete(key)
+        elif dkey is not None and dvalue is not None:
+            for key in app.kvsession_store.keys():
+                if app.session_interface.serialization_method.loads(
```

Comment:
Why do you guard against ValueError in L115 and not here?

Reply:
L115 is a SessionID.unserialize call and this does not handle "invalid" session cookies very well.
L127 unpickles data from redis - this should not fail unless redis flipped some bits.
If it fails it raises a _pickle.UnpicklingError.